File: java/HeapPackage/MinHeap.java
package HeapPackage;
public class MinHeap {
    private int[] Heap;
    private int size;
    private int maxsize;

    public MinHeap(int maxsize) {
        this.maxsize = maxsize;
        this.size = 0;
        Heap = new int[this.maxsize + 1];
        Heap[0] = Integer.MIN_VALUE;
    }
    public int getSize() {
        return size;
    }

    private int parent(int pos) {
        return pos / 2;
    }

    private int leftChild(int pos) {
        return (2 * pos);
    }

    private int rightChild(int pos) {
        return (2 * pos) + 1;
    }

    private boolean isLeaf(int pos) {
        if (pos > (size / 2) && pos <= size) {
            return true;
        }
        return false;
    }

    private void swap(int fpos, int spos) {
        int tmp;
        tmp = Heap[fpos];
        Heap[fpos] = Heap[spos];
        Heap[spos] = tmp;
    }

    private void minHeapify(int pos) {
        if (!isLeaf(pos)) {
            int left = leftChild(pos);
            int right = rightChild(pos);
    
            // Check if current node is larger than its children
            if ((left < size && Heap[pos] > Heap[left]) || 
                (right < size && Heap[pos] > Heap[right])) {
                // Swap with the smaller child and continue heapifying
                if (right >= size || Heap[left] < Heap[right]) {
                    swap(pos, left);
                    minHeapify(left);
                } else {
                    swap(pos, right);
                    minHeapify(right);
                }
            }
        }
    }
    

    public void insert(int element) {
        if (size >= maxsize) {
            return;
        }
        Heap[++size] = element;
        int current = size;

        while (Heap[current] < Heap[parent(current)]) {
            swap(current, parent(current));
            current = parent(current);
        }
    }

    public int remove() {
        int popped = Heap[1];
        Heap[1] = Heap[size--];
        minHeapify(1);
        return popped;
    }
}

File: cpp/min_heap.cpp
#include <iostream>
#include <vector>

class MinHeap {
private:
    std::vector<int> heap;

    int parent(int i) { return (i - 1) / 2; }
    int left(int i) { return (2 * i + 1); }
    int right(int i) { return (2 * i + 2); }

    void heapifyDown(int i) {
        int leftChild = left(i);
        int rightChild = right(i);
        int smallest = i;

        if (leftChild < heap.size() && heap[leftChild] < heap[i])
            smallest = leftChild;
        if (rightChild < heap.size() && heap[rightChild] < heap[smallest])
            smallest = rightChild;

        if (smallest != i) {
            std::swap(heap[i], heap[smallest]);
            heapifyDown(smallest);
        }
    }

    void heapifyUp(int i) {
        if (i && heap[parent(i)] > heap[i]) {
            std::swap(heap[i], heap[parent(i)]);
            heapifyUp(parent(i));
        }
    }

public:
    void insert(int key) {
        heap.push_back(key);
        int index = heap.size() - 1;
        heapifyUp(index);
    }

    int extractMin() {
        if (heap.size() == 0)
            return INT_MAX;

        int root = heap.front();
        heap[0] = heap.back();
        heap.pop_back();

        heapifyDown(0);
        return root;
    }

    void printHeap() {
        for (int i = 0; i < heap.size(); ++i)
            std::cout << heap[i] << " ";
        std::cout << "\n";
    }
};
